# Credit Risk Ensemble Techniques

In [1]:
import warnings
warnings.filterwarnings('ignore')

In [2]:
# Initial imports
import numpy as np
import pandas as pd
from pathlib import Path
from collections import Counter

In [3]:
from sklearn.metrics import balanced_accuracy_score
from sklearn.metrics import confusion_matrix, accuracy_score, classification_report
from imblearn.metrics import classification_report_imbalanced
from sklearn.model_selection import train_test_split
#from sklearn.preprocessing import StandardScaler
from imblearn.ensemble import BalancedRandomForestClassifier
from imblearn.ensemble import EasyEnsembleClassifier 






# Read the CSV and Perform Basic Data Cleaning

In [4]:
# Load the data
file_path = Path('Resources/lending_data.csv')
df = pd.read_csv(file_path)

df.head()

loan_size  interest_rate homeowner  borrower_income  debt_to_income  \
0    10700.0          7.672       own            52800        0.431818   
1     8400.0          6.692       own            43600        0.311927   
2     9000.0          6.963      rent            46100        0.349241   
3    10700.0          7.664       own            52700        0.430740   
4    10800.0          7.698  mortgage            53000        0.433962   

   num_of_accounts  derogatory_marks  total_debt loan_status  
0                5                 1       22800    low_risk  
1                3                 0       13600    low_risk  
2                3                 0       16100    low_risk  
3                5                 1       22700    low_risk  
4                5                 1       23000    low_risk

In [5]:
df.isnull().values.any()

False

# Split the Data into Training and Testing

In [6]:
# Create our features
X = df.copy()
X.drop(columns=["loan_status"], axis=1, inplace=True)
X.head()

loan_size  interest_rate homeowner  borrower_income  debt_to_income  \
0    10700.0          7.672       own            52800        0.431818   
1     8400.0          6.692       own            43600        0.311927   
2     9000.0          6.963      rent            46100        0.349241   
3    10700.0          7.664       own            52700        0.430740   
4    10800.0          7.698  mortgage            53000        0.433962   

   num_of_accounts  derogatory_marks  total_debt  
0                5                 1       22800  
1                3                 0       13600  
2                3                 0       16100  
3                5                 1       22700  
4                5                 1       23000

In [7]:
# Use `get_dummies` to generate the features for homeowner
X = pd.get_dummies(X, columns=['homeowner']).copy()
X.head()

loan_size  interest_rate  borrower_income  debt_to_income  num_of_accounts  \
0    10700.0          7.672            52800        0.431818                5   
1     8400.0          6.692            43600        0.311927                3   
2     9000.0          6.963            46100        0.349241                3   
3    10700.0          7.664            52700        0.430740                5   
4    10800.0          7.698            53000        0.433962                5   

   derogatory_marks  total_debt  homeowner_mortgage  homeowner_own  \
0                 1       22800                   0              1   
1                 0       13600                   0              1   
2                 0       16100                   0              0   
3                 1       22700                   0              1   
4                 1       23000                   1              0   

   homeowner_rent  
0               0  
1               0  
2               1  
3               0  
4               0

In [8]:
# Create our target
y = df["loan_status"]
y.describe()


count        77536
unique           2
top       low_risk
freq         75036
Name: loan_status, dtype: object

In [9]:
X.describe()

loan_size  interest_rate  borrower_income  debt_to_income  \
count  77536.000000   77536.000000     77536.000000    77536.000000   
mean    9805.562577       7.292333     49221.949804        0.377318   
std     2093.223153       0.889495      8371.635077        0.081519   
min     5000.000000       5.250000     30000.000000        0.000000   
25%     8700.000000       6.825000     44800.000000        0.330357   
50%     9500.000000       7.172000     48100.000000        0.376299   
75%    10400.000000       7.528000     51400.000000        0.416342   
max    23800.000000      13.235000    105200.000000        0.714829   

       num_of_accounts  derogatory_marks    total_debt  homeowner_mortgage  \
count     77536.000000      77536.000000  77536.000000        77536.000000   
mean          3.826610          0.392308  19221.949804            0.497472   
std           1.904426          0.582086   8371.635077            0.499997   
min           0.000000          0.000000      0.000000            0.000000   
25%           3.000000          0.000000  14800.000000            0.000000   
50%           4.000000          0.000000  18100.000000            0.000000   
75%           4.000000          1.000000  21400.000000            1.000000   
max          16.000000          3.000000  75200.000000            1.000000   

       homeowner_own  homeowner_rent  
count   77536.000000    77536.000000  
mean        0.398911        0.103616  
std         0.489678        0.304764  
min         0.000000        0.000000  
25%         0.000000        0.000000  
50%         0.000000        0.000000  
75%         1.000000        0.000000  
max         1.000000        1.000000

In [10]:
# Check the balance of our target values
y.value_counts()

low_risk     75036
high_risk     2500
Name: loan_status, dtype: int64

In [11]:
# Split the X and y into X_train, X_test, y_train, y_test
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=1, stratify=y)
X_train.shape


(58152, 10)

The code below was initially ran to compare the output of the BAC if scaler was applied vs if not applied. However, the BAC was slightly higher when the scaler was not applied. Thus the code is commented out and not incorporated in the output

In [12]:
#Use the StandardScaler to scale the features data - X_train
#Creating StandardScaler instance
# from imblearn.ensemble import EasyEnsembleClassifier 
# scaler = StandardScaler()

In [13]:
# Fitting Standard Scaller
# X_scaler = scaler.fit(X_train)

In [14]:
# Scaling data
# X_train_scaled = X_scaler.transform(X_train)
# X_test_scaled = X_scaler.transform(X_test)

# Ensemble Learners

In this section, you will compare two ensemble algorithms to determine which algorithm results in the best performance. You will train a Balanced Random Forest Classifier and an Easy Ensemble classifier . For each algorithm, be sure to complete the folliowing steps:

1. Train the model using the training data. 
2. Calculate the balanced accuracy score from sklearn.metrics.
3. Print the confusion matrix from sklearn.metrics.
4. Generate a classication report using the `imbalanced_classification_report` from imbalanced-learn.
5. For the Balanced Random Forest Classifier onely, print the feature importance sorted in descending order (most important feature to least important) along with the feature score

Note: Use a random state of 1 for each algorithm to ensure consistency between tests

### Balanced Random Forest Classifier

In [15]:
# Resample the training data with the BalancedRandomForestClassifier

rf_model = BalancedRandomForestClassifier(n_estimators=100, random_state=1)
rf_model.fit(X_train, y_train)
y_pred = rf_model.predict(X_test)

In [16]:
# Calculated the balanced accuracy score
acc_score = accuracy_score(y_test, y_pred)
print(f"Balanced Accuracy Score : {acc_score}")

Balanced Accuracy Score : 0.9922616591002889


In [17]:
# Calculating the confusion matrix
confusionmatrix = confusion_matrix(y_test, y_pred)
cm_df = pd.DataFrame(
    confusionmatrix, index=["Actual 0", "Actual 1"], columns=["Predicted 0", "Predicted 1"]
)

In [18]:
# Displaying results
print("Confusion Matrix")
print("----------------------------------")
display(cm_df)


Confusion Matrix
----------------------------------


Predicted 0  Predicted 1
Actual 0          622            3
Actual 1          147        18612

In [19]:
# Print the imbalanced classification report
print(classification_report_imbalanced(y_test, y_pred))

                   pre       rec       spe        f1       geo       iba       sup

  high_risk       0.81      1.00      0.99      0.89      0.99      0.99       625
   low_risk       1.00      0.99      1.00      1.00      0.99      0.99     18759

avg / total       0.99      0.99      1.00      0.99      0.99      0.99     19384



In [20]:
# List the features sorted in descending order by feature importance
importances = rf_model.feature_importances_
sorted(zip(importances, X.columns), reverse=True)

[(0.22759633831687995, 'borrower_income'),
 (0.1965574723230606, 'interest_rate'),
 (0.17417807060091195, 'debt_to_income'),
 (0.14944360428565667, 'total_debt'),
 (0.14229163194622926, 'loan_size'),
 (0.07859627283289125, 'num_of_accounts'),
 (0.02885770833789564, 'derogatory_marks'),
 (0.001219411955649641, 'homeowner_own'),
 (0.0008974701883394341, 'homeowner_mortgage'),
 (0.0003620192124855904, 'homeowner_rent')]

### Easy Ensemble Classifier

In [21]:
# Train the EasyEnsembleClassifier
eec = EasyEnsembleClassifier(random_state=1, n_estimators=100)
eec.fit(X_train, y_train) 

EasyEnsembleClassifier(n_estimators=100, random_state=1)

In [22]:
# Predict 
y_pred = eec.predict(X_test)
print(confusion_matrix(y_test, y_pred))

[[  622     3]
 [  118 18641]]


In [23]:
# Calculated the balanced accuracy score
acc_score = accuracy_score(y_test, y_pred)
print(f"Balanced Accuracy Score : {acc_score}")

Balanced Accuracy Score : 0.9937577383408998


In [24]:
# Display the confusion matrix
# Calculating the confusion matrix
cm = confusion_matrix(y_test, y_pred)
cm_df = pd.DataFrame(
    cm, index=["Actual 0", "Actual 1"], columns=["Predicted 0", "Predicted 1"]
)

In [25]:
# Displaying results
print("Confusion Matrix")
print("----------------------------------")
display(cm_df)

Confusion Matrix
----------------------------------


Predicted 0  Predicted 1
Actual 0          622            3
Actual 1          118        18641

In [26]:
# Print the imbalanced classification report
print(classification_report_imbalanced(y_test, y_pred))

                   pre       rec       spe        f1       geo       iba       sup

  high_risk       0.84      1.00      0.99      0.91      0.99      0.99       625
   low_risk       1.00      0.99      1.00      1.00      0.99      0.99     18759

avg / total       0.99      0.99      1.00      0.99      0.99      0.99     19384



Analysis:

Model with the best balanced accuracy score?
The easy ensemble classifier with 99.4%

Model with the best recall score?
Both models had the same score of 1.0 for high risk

Model with the best geometric mean score?
The geometrical mean was the same at 0.99

The top three features?
Borrowers income (22.76%), Interest rate (19.66%) and debt to income (17.42%)